scripts/lint_jjb.py

```diff
@@ -176,144 +181,230 @@ def check_retention(job, in_file):
+        # replace a hash range with the first value in the
+        # range
+        schedule = re.sub(r'H\((\d+)-\d+\)', r'\1', schedule)
 
-    cron_value = data.get('CRON')
+    # then handle the H pattern
+    schedule = re.sub(r'H', rep_char, schedule)
 
-    return_value = 0
-    if check_timed_trigger_value(cron_value) == 1:
-        sys.stderr.write(error_message)
-        return_value = 1
+    return schedule
 
-    return return_value
 
+def is_valid_cron_statement(schedule, name):
+    raw_schedule = schedule
+    schedule = translate_hash(schedule)
+    if croniter.is_valid(schedule):
+        return True
+    else:
+        sys.stderr.write("{n} Invalid cron expression: {s}"
+                         " (translated from {rs})\n"
+                         .format(n=name, s=schedule, rs=raw_schedule))
+        return False
 
-def check_timed_trigger_cron_std(data, in_file):
 
+def is_comment(schedule):
+    if re.match(r'^\s*#', schedule):
+        return True
+    else:
+        return False
+
+
+def allowed_stdjob_schedule(build_schedule, data):
+    """ Standard jobs should run at most daily"""
+    # non standard jobs are not validated by this rule
+    if not is_standard_job(data):
+        return True
+
+    raw_schedule = build_schedule
+    # This check is about the interval, so the H values
+    # in the input cron expression should be translated
+    # to a single value as Jenkins does
+    # rather than * which represents all values.
+    build_schedule = translate_hash(build_schedule, rep_mode="one")
+    name = data['name']
+
+    base = datetime.now()
+    build_iter = croniter(build_schedule, base)
+    build_start_1 = build_iter.get_next(datetime)
+    build_start_2 = build_iter.get_next(datetime)
+    delta = build_start_2 - build_start_1
+    if delta < timedelta(hours=23):
+        sys.stderr.write(
+            "{n} Is scheduled too frequently (every {delta}), standard"
+            " jobs should be executed at most daily. Cron: {rs},"
+            " translated to: {s}\n".format(
+                n=name,
+                delta=delta,
+                rs=raw_schedule,
+                s=build_schedule))
+        return False
+    return True
+
+
+def no_maint_window_conflict(build_schedule, name):
+    raw_schedule = build_schedule
+
+    # Convert H --> * as Jenkins could assign
+    # any value to H.
+    build_schedule = translate_hash(build_schedule)
+
+    base = datetime.now()
+    max_build_duration = timedelta(hours=10)
+
+    maint_schedule = "0 10 * * 1"
+    maint_iter = croniter(maint_schedule, base)
+    maint_duration = timedelta(hours=2)
+
+    # Find the interval between builds
+    build_iter = croniter(build_schedule, base)
+    next_build = build_iter.get_next(datetime)
+    second_build = build_iter.get_next(datetime)
+    build_interval = second_build - next_build
+
+    # check the next 60 maintenance windows for conflicts
+    # this is over a year so should cover most eventualities.
```

Comment:
The variable in this case is the day of the month field. Maybe instead you could put some constraints around that field and then you don't need to be concerned about it varying with time.